src/librustc/hir/lowering/item.rs

```diff
@@ -1219,8 +1219,44 @@ impl LoweringContext<'_> {
             let async_expr = this.make_async_expr(
                 CaptureBy::Value, closure_id, None, body.span,
                 |this| {
-                    let body = this.lower_block_with_stmts(body, false, statements);
-                    this.expr_block(body, ThinVec::new())
+                    // Create a block from the user's function body:
+                    let user_body = this.lower_block(body, false);
+                    let user_body = this.expr_block(user_body, ThinVec::new());
```

Comment:
This `lower_block` + `expr_block` happens in the lowering of:

- `ExprKind::Async`
- `lower_expr_if`
- `lower_expr_while_in_loop_scope`

Introduce `fn lower_block_expr` and refactor those?